File: 240616/금 채굴하기/gold-mining.py
from collections import deque

N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]

def get_max_golds(sy, sx, n, m, board):
    max_golds = 0
    visited = [[False] * n for _ in range(n)]
    f_dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    for k in range(n):
        for i in range(n):
            visited[i] = [False] * n
        
        visited[sy][sx] = True
        BFS = deque([(sy, sx, 0)])
        total_golds = 0

        while BFS:
            y, x, side = BFS.popleft()

            if k < side:
                break

            total_golds += board[y][x]
            for dy, dx in f_dirs:
                my, mx = y + dy, x + dx

                if my < 0 or mx < 0 or my >= n or mx >= n:
                    continue

                if visited[my][mx]:
                    continue

                visited[my][mx] = True
                BFS.append((my, mx, side + 1))

        cost = k * k + (k + 1) * (k + 1)
        if cost <= total_golds * m:
            max_golds = max(max_golds, total_golds)

    return max_golds

max_golds = 0
for y in range(N):
    for x in range(N):
        max_golds = max(max_golds, get_max_golds(y, x, N, M, board))

print(max_golds)

File: 240616/금 채굴하기/gold-mining.cpp
#include <iostream>
#include <queue>
#include <tuple>

int n, gold_price;

int board[20][20];
int visited[20][20];

int dy[4] = { -1, 0, 1, 0 };
int dx[4] = { 0, 1, 0, -1 };

int answer;

void GetMaxGold(const int start_y, const int start_x)
{
       // 최대 시도해볼 수 있는 곳
	for (int k = 0; k < n; ++k)
	{
		for (int i = 0; i < n; ++i)
			std::fill(visited[i], visited[i] + n, false);

		visited[start_y][start_x] = true;
		std::queue<std::tuple<int, int, int>> bfs;
		bfs.push({ start_y, start_x, 0 });

		int total_gold_acquired = 0;
		while (bfs.size())
		{
			int y, x, step;
			std::tie(y, x, step) = bfs.front();
			bfs.pop();

			if (k < step)
				break;

			total_gold_acquired += board[y][x];

			for (int i = 0; i < 4; ++i)
			{
				int ny = y + dy[i];
				int nx = x + dx[i];

				if (ny < 0 || nx < 0 || n <= ny || n <= nx)
					continue;

				if (visited[ny][nx])
					continue;

				visited[ny][nx] = true;
				bfs.push({ ny, nx, step + 1 });
			}
		}

		int cost = k * k + (k + 1) * (k + 1);
		if (cost <= total_gold_acquired * gold_price)
			answer = std::max(answer, total_gold_acquired);
	}
}

int main()
{
	// 여기에 코드를 작성해주세요.
	std::cin >> n >> gold_price;
	for (int y = 0; y < n; ++y)
	{
		for (int x = 0; x < n; ++x)
			std::cin >> board[y][x];
	}

	for (int y = 0; y < n; ++y)
	{
		for (int x = 0; x < n; ++x)
			GetMaxGold(y, x);
	}

	std::cout << answer;

	return 0;
}